packages/rest/src/writer.ts

```diff
@@ -12,21 +12,37 @@ import {Readable} from 'stream';
  *
  * @param response HTTP Response
  * @param result Result from the API to write into HTTP Response
+ * @param contentType Optional content type for the response
  */
 export function writeResultToResponse(
   // not needed and responsibility should be in the sequence.send
   response: Response,
   // result returned back from invoking controller method
   result: OperationRetval,
+  // content type for the response
+  contentType?: string,
 ): void {
-  if (!result) {
-    response.statusCode = 204;
+  // Check if response writing should be bypassed. When the return value
+  // is the response object itself, no writing should happen.
+  if (result === response) return;
+
+  if (result === undefined) {
+    response.status(204);
     response.end();
     return;
   }
 
-  if (result instanceof Readable || typeof result.pipe === 'function') {
-    response.setHeader('Content-Type', 'application/octet-stream');
+  function setContentType(defaultType: string = 'application/json') {
+    if (response.getHeader('Content-Type') == null) {
```

Comment:
Personally, I prefer to revert this change.

Ideally, there should be only two ways how a content-type is determined:

1. The controller method/route returns data to be serialized to the response. The content-type is determined by the framework depending on the type of data (object vs. string vs. stream or buffer), request headers like `Accepts` and endpoint OpenAPI spec. Right now, we have a very limited version of this algorithm implemented in the current `writeResultToResponse` code (objects are sent as JSON, streams as binary, strings as text).

2. The controller wants to set an explicit content type, status code or change any other aspect of the HTTP response. It returns an object describing response properties (status code, headers, body) and the `send` action applies these instructions to the `response` object. This will be enabled by #436 in the future.

There is of course also the fourth mode being added here, I think it should be the only scope of this pull request,  and I consider it as a short-term workaround:

4. The controller (or a sequence action in general) takes full control of the response sent: it sets the status code & headers and writes the response body. The `send` action becomes a no-op.